Have set() and set_quoted() of configuration object work with newlines.

Jehan opened the pull request:
set_quoted() makes the value into a double-quoted string, so let's
assume C-style string, in particular with newlines as "\n".

Also take care of remaining newlines in dump_conf_header(). C or nasm
macros expect single-line values so if the value was multi-line, we
would end up with broken syntax. Appending a backslash at each end of
line make them concat into a single line in both C and nasm format
(note: multi-line macros in nasm are actually possible apparently but
use another format not outputted by current meson code). Also note that
the replacement is done at the end only when dumping the conf as a
header because we cannot assume anything about the format when replacing
variables from an input file (in this case, it should be the dev
responsibility).

----------

This was a result of discovering broken files when setting multi-line values to a conf variable:
```
conf.set('VAR', 'something
with newlines')
```
Of course my example seems weird because in real life, the value was simply the output of some command (which was multi-line and that was exactly what I wanted, not a bug).

Then when creating a `config.h`, it ended up broken (hence broke the build):
```
#define VAR something
with newlines
```

The solution is to append backslashes:
```
#define VAR something\
with newlines
```

Now actually I was using a set_quoted():
```
conf.set_quoted('VAR', 'something
with newlines')
```
In such case, it seems obvious we want to use this like a string, so let's replace newlines with `\n` directly to make sure this works well. In particular, we want to get:
```
#define VAR something\nwith newlines
```

This commit deals with all these cases.

jpakkane commented:
The configure file test needs to be extended to set a variable with an embedded newline and then verifying that it is written to the output file in the correct format.

Jehan commented:
Ok so I fixed a bug in the previous commit (amended it directly) and added 2 tests (for set() and set_quoted() respectively) in a new commit. Hope this is good now! :-)

jpakkane commented on test cases/common/14 configure file/meson.build:
```diff
@@ -88,6 +88,9 @@ dump.set('SHOULD_BE_ONE', 1)
 dump.set('SHOULD_BE_QUOTED_ONE', '"1"')
 
+dump.set('MULTILINE', '1\n+2')
+dump.set_quoted('MULTILINE_STRING', 'line1\nline2')
```
Use a Meson multiline string here:

```meson
dump.set_quoted('MULTILINE_STRING', '''line1
line2''')
```

Jehan replied:
Done.

jpakkane commented:
Thanks.

xclaesse commented:
This breaks build of glib, it adds extra \ at the end of many lines in its config.h

xclaesse commented:
Reverted because it's not a backward compatible change. You have to add a new kwarg to allow multiline.

eli-schwartz commented:
> Reverted because it's not a backward compatible change.

And any change that prevents an SQL injection vulnerability also breaks backwards compat if people deliberately abused an injection attack.

This is an injection attack on the conf file format, violating the semantics of the concept of set(). It should not be allowed, and glib should be responsible for fixing what they broke by deliberately attempting to abuse what they knew was a bug.
